backend/src/composite/mod.rs

```diff
@@ -247,52 +396,88 @@ impl<'a, F: FieldElement> Backend<'a, F> for CompositeBackend<'a, F> {
-                        })
-                    }
+        // We use scoped threads to be able to share non 'static references.
+        thread::scope(|scope| {
+            let mut proofs_status = self
+                .machine_data
+                .iter()
+                .map(|machine_entry| {
+                    let (machine, machine_data) = machine_entry;
+                    let (witness, size) =
+                        process_witness_for_machine(machine, machine_data, witness);
+
+                    let inner_machine_data = machine_data
+                        .get(&size)
+                        .expect("Machine does not support the given size");
+
+                    let status = time_stage(machine, size, 0, || {
+                        sub_prover::run(scope, &inner_machine_data.backend, witness)
+                    });
+
+                    (status, machine_entry, size)
+                })
+                .collect::<Vec<_>>();
+
+            let mut proof_results = Vec::new();
+            let mut witness = Cow::Borrowed(witness);
+            for stage in 1.. {
+                // Filter out proofs that have completed and accumulate the
+                // challenges.
+                let mut challenges = BTreeMap::new();
+                let waiting_provers = std::mem::take(&mut proofs_status)
+                    .into_iter()
+                    .filter_map(|(status, machine_data, size)| match status {
+                        sub_prover::RunStatus::Completed(result) => {
+                            proof_results.push((result, size));
+                            None
+                        }
+                        sub_prover::RunStatus::Challenged(sub_prover, c) => {
+                            // Accumulate the challenges
+                            accumulate_challenges(&mut challenges, c);
+                            Some((sub_prover, machine_data))
+                        }
+                    })
+                    .collect::<Vec<_>>();
+
+                if waiting_provers.is_empty() {
+                    break;
+                }
+
+                witness = witgen_callback
+                    .next_stage_witness(&witness, challenges, stage)
+                    .into();
+
+                // Resume the waiting provers with the new witness
+                proofs_status = waiting_provers
+                    .into_iter()
+                    .map(|(prover, machine_entry)| {
+                        let (machine_name, machine_data) = machine_entry;
+                        let (witness, size) =
+                            process_witness_for_machine(machine_name, machine_data, &witness);
+
+                        let status = time_stage(machine_name, size, stage + 1, move || {
+                            prover.resume(witness)
+                        });
+
+                        (status, machine_entry, size)
+                    })
+                    .collect();
```

Comment:
Not sure if it's worth it, but I think if the `run` function didn't immediately `wait()`, we could could simplify this a bit, like this:

- Call `run()` for each machine. The threads should block immediately.
- `for stage in 0..`:
  - If `stage != 0`: Update witness by calling `witgen_callback.next_stage_witness`
  - Split witness and resume (this very block)
  - Check proof status and append proof or add to `waiting_provers`
  - Abort if `waiting_provers` is empty.

This would get rid of most code before this loop in the `thread::scope` call.

Reply:
It seems to me that the only thing that will be saved in the code before the loop is the `process_witness_for_machine` call... but let me try.

Reply:
Ok, tried, but I couldn't do it, because I need to call `process_witness_for_machine` before calling `run` in order to get the adequately sized backend (i.e. `machine_data[&size].backend`).

Reply:
OK, thanks for checking!